### DISCUSSION TITLE
_use_snes_vi_solver and _barrier

### DISCUSSION THREAD

**Discussion by selarem** [idaholab/moose#26534] [Q&A Modules: Phase field]
Hello,
When considering  use_snes_vi_solver in  ComputeLinearElasticPFFractureStress.C, I think that it does not work properly when we do have _barrier : the damage is negatif at the biginning of the simulation until barrier reached.  The block that assigns history variable is :
``` 
// // Assign history variable
  Real hist_variable = _H_old[_qp];
  if (_use_snes_vi_solver)
  {
    _H[_qp] = F_pos;

    if (_use_current_hist)
      hist_variable = _H[_qp];
  }
  else
  {
    if (F_pos > _H_old[_qp])
      _H[_qp] = F_pos;
    else
      _H[_qp] = _H_old[_qp];

    if (_use_current_hist)
      hist_variable = _H[_qp];
 
    if (hist_variable < _barrier[_qp])
     hist_variable = _barrier[_qp];
  }  
```

I think that the two last instructions should be put outside like this:
```
// // Assign history variable
  Real hist_variable = _H_old[_qp];
  if (_use_snes_vi_solver)
  {
    _H[_qp] = F_pos;

    if (_use_current_hist)
      hist_variable = _H[_qp];
  }
  else
  {
    if (F_pos > _H_old[_qp])
      _H[_qp] = F_pos;
    else
      _H[_qp] = _H_old[_qp];

    if (_use_current_hist)
      hist_variable = _H[_qp];
// Original
//    if (hist_variable < _barrier[_qp])
 //     hist_variable = _barrier[_qp];
  }  
  //  Correction 11/01/2024
  if (hist_variable < _barrier[_qp])
                hist_variable = _barrier[_qp];
```
regards,
saber

**Comment by GiudGiud** [comment]
My guess is that the SNES VI solver should be imposing that constraining so we should not need to do it ourselves?
but @jiangwen84  will know for sure

**Comment by jiangwen84** [comment]
@selarem your changes seem to be reasonable to me. When we initially implemented it, we only considered either using VI without barrier or using barrier without VI, did not think about combing these two. 

Are you willing to make a PR to correct it? Please do not forget to add a regression test.

**Reply by selarem** [reply]
There's another problem that we have observed concerning the bandwidth of the damaged zone: when a barrier is considered in the case of linear fracture energy (3/8 *  gc /Lc  * d) the band width is theoretically 4 * Lc (Lc being the internal  length scale). This will never be obtained in the actual implementation in moose!  The instruction : hist_variable = _barrier[_qp]; should never be executed means invoked. This used to happen (before our correction suggestion) when _use_snes_vi_solver = true (and the correct bounds for the Executioner block).  
So in summary, the older version was fine in the unique case of    considering a barrier and using snes solver.  When  _use_snes_vi_solver = false;  the bandwidth could be 20*Lc. 
with the correction we have suggested, whenever we have a barrier (with or without use of the snes solver) , the bandwidth is too large. 

So in case of linear fracture energy, the snes solver has to be mandatory.  So the code should be something like this:


```
 if (_barrier)
_use_snes_vi_solver = true; 

if (_use_snes_vi_solver)
  {
    _H[_qp] = F_pos;

    if (_use_current_hist)
      hist_variable = _H[_qp];
  } 

```
